feat: new options

AleksanderBodurri opened the pull request:
Includes new options for the following:

```
lowerBound: number;
applyMutations: (node: Node) => void
```

Options are documented in README

`applyMutations` could be used to modify node colors. Therefore this PR solves part of #5 . I believe it could also be used to solve #28 , either with javascript or by setting the `title` attribute on the node DOM element.
 
Added in useful gitignore targets aswell.

evmar commented on src/treemap.ts:
```diff
@@ -55,11 +50,19 @@ export function isDOMNode(e: Element): boolean {
+  lowerBound: number;
+  classList: ((node: Node) => string)[];
+  nodeColor: (node: Node) => NodeColor;
+  caption(node: Node): string;
```
Did you mean to make this non-optional?

AleksanderBodurri replied:
Yes. As far as i can tell everywhere an `options` variable is typed it is done so as `Partial<Options>` which makes the optional field redundant.

evmar commented on src/treemap.ts:
```diff
@@ -55,11 +50,19 @@ export function isDOMNode(e: Element): boolean {
 export interface Options {
   padding: [number, number, number, number];
-  caption?(node: Node): string;
+  lowerBound: number;
```
Can you doc this somewhere?  Maybe right here as a jsdoc comment?

evmar commented on src/treemap.ts:
```diff
@@ -55,11 +50,19 @@ export function isDOMNode(e: Element): boolean {
+interface NodeColor {
+  defaultColor: string;
+  hoverColor: string;
+}
```
I wonder, can we achieve this with CSS?

AleksanderBodurri replied:
I wondered the same thing. My initial attempt to do it this way was how the `classList` option was born. 

Because `:hover` is a pseudo selector, it can't be changed with any DOM APIs. To achieve this with css you would have to inject style rules programatically into the page and have them specifically target certain nodes based on the callback you pass in.

Imagine a set of data where we want the node with the greatest size to be red and every other node a lighter shade of red proportional to the ratio of its size to the largest node's size.

I could not think of a css solution to this case so I decided to go with the `mouseleave` and `mouseenter` events to edit the style of each DOM element inline. I would love to discuss alternate solutions if you have any in mind.

evmar replied:
Wow, it is disappointing that you can't do this with inline style!
Is there some way we can leave it up to the caller to do CSS tweaks other than the hover?  E.g. via the classlist thing?

AleksanderBodurri replied:
I've had a moment to reflect on this and agree that the current solution is not ideal. The use case I describe could potentially be solved on the user end using more advanced css (scss @for loops for example) and the classList option. I appreciate you calling this out, sometimes these things have to be discussed for an ideal solution to be reached.

I'll remove this option in favour of the user defined class approach. I suspect this would greatly improve performance as well for larger trees.

evmar commented on src/treemap.ts:
```diff
@@ -102,6 +105,9 @@ function px(x: number): string {
 function defaultOptions(options: Partial<Options>): Options {
   const opts = {
     padding: options.padding || [14, 3, 3, 3],
+    lowerBound: options.lowerBound === undefined ? 0.1 : options.lowerBound,
```
Is this the same:
options.lowerBound ?? 0.1

AleksanderBodurri replied:
Yes it is. If you want I can attempt to upgrade the ts version to 3.7+ but when I attempted it originally there were other errors being thrown by the ts compiler. I think that a separate PR for a ts version bump is more appropriate.

evmar replied:
Oh yeah, sorry!  My mind is in my work where we're already using 3.7 :)

evmar commented on src/treemap.ts:
```diff
@@ -128,6 +134,8 @@ export class TreeMap {
     const dom = document.createElement('div');
     dom.className = NODE_CSS_CLASS;
+    this.options.classList.forEach(nodeClass => dom.classList.add(nodeClass(node)));
+    this.setDOMColour(dom, this.options.nodeColor(node));
```
I think maybe better to consistently use American spelling.  (This line has both colour and color in it.)

AleksanderBodurri replied:
Whoops, I meant to change all those to the American spelling, missed this one. Thanks for the catch.

AleksanderBodurri commented:
> This is great, thanks for writing it!
> 
> I think it would be good document what all these new options do, either here or in the README.

Thank you for the very useful library 👊

I added one more option for font coloring and added documentation for all existing options to the readme. Should be good to go whenever you're ready to review @evmar

evmar commented on README.md:
```diff
@@ -33,6 +33,37 @@ const container = document.getElementById('myContainer');
+| Option     | Description      |
+| ------------- |:-------------:|
+| padding      | In order: padding-top, padding-right, padding-bottom, padding-left of each node
+| lowerBound   | Lower bound of ratio that determines how many children can be displayed inside of a node. Example with a lower bound of 0.1: the total area taken up by displaying child nodes of any given node cannot be less than 10% of the area of it's parent node. 
```
typo: it's => its

evmar commented on src/treemap.ts:
```diff
@@ -128,6 +135,8 @@ export class TreeMap {
     if (node.dom) return node.dom;
     const dom = document.createElement('div');
     dom.className = NODE_CSS_CLASS;
+    this.options.classList.forEach(nodeClass => dom.classList.add(nodeClass(node)));
```
This API doesn't make sense to me.  Why accept an array of functions, rather than just one?  What's the use case?

Maybe a single function that can return an array makes more sense?
Maybe we accept an option that lets us just pass each node to the caller to let them mutate as they will, modifying classList or anything else?

AleksanderBodurri replied:
I like the idea of having a `mutations` option, it could solve a variety of use cases. Great call, will replace this.

AleksanderBodurri commented:
> Thank you, the docs look great. Also sorry for my slow response.

No worries, I'll get to work on the changes as soon as I get the chance.

AleksanderBodurri commented:
Changes made. Let me know what you think :) @evmar

AleksanderBodurri commented:
Hey @evmar any updates on this?

evmar commented:
Thank you!  (Also thank you for pinging me again, I am so easily distracted.)
